backbone.js

```diff
@@ -106,46 +95,50 @@
     // callbacks for the event. If `name` is null, removes all bound
     // callbacks for all events.
     off: function(name, callback, context) {
-      if (!this._events || !eventsApi(this, 'off', name, [callback, context])) return this;
+      if (!this._events || !eventsApi(this, this.off, name, [callback, context])) return this;
 
       // Remove all callbacks for all events.
       if (!name && !callback && !context) {
-        this._events = void 0;
+        var listenerIds = _.keys(this._listeners);
+        for (var i = 0, l = listenerIds.length; i < l; i++) {
+          var listener = this._listeners[listenerIds[i]].obj;
+          delete listener._listeningTo[this._listenId];
+        }
+        this._events = this._listeners = void 0;
         return this;
       }
 
       var names = name ? [name] : _.keys(this._events);
       for (var i = 0, length = names.length; i < length; i++) {
-        name = names[i];
+        var eventName = names[i];
 
         // Bail out if there are no events stored.
-        var events = this._events[name];
+        var events = this._events[eventName];
         if (!events) continue;
 
-        // Remove all callbacks for this event.
-        if (!callback && !context) {
-          delete this._events[name];
-          continue;
-        }
-
         // Find any remaining events.
         var remaining = [];
         for (var j = 0, k = events.length; j < k; j++) {
           var event = events[j];
+          var listener = this._listeners && this._listeners[event.listenId];
           if (
             callback && callback !== event.callback &&
             callback !== event.callback._callback ||
-            context && context !== event.context
+            context && context !== event.context ||
+            name && name !== eventName
```

Comment:
This `name && name !== eventName` condition should be taken care of by [L111](https://github.com/jashkenas/backbone/pull/3049/files#diff-0d56d0d310de7ff18b3cef9c2f8f75dcR111)